packages/adapters-library/src/adapters/etherfi/products/l2/etherFiL2Adapter.ts

```diff
@@ -0,0 +1,227 @@
+      },
+    ]
+  }
+
+  private async getProtocolTokenByAddress(
+    protocolTokenAddress: string,
+  ): Promise<ProtocolToken> {
+    return this.helpers.getProtocolTokenByAddress({
+      protocolTokens: await this.getProtocolTokens(),
+      protocolTokenAddress,
+    })
+  }
+
+  async getPositions(input: GetPositionsInput): Promise<ProtocolPosition[]> {
+    return this.helpers.getBalanceOfTokens({
+      ...input,
+      protocolTokens: await this.getProtocolTokens(),
+    })
+  }
+
+  async getWithdrawals({
+    protocolTokenAddress,
+    fromBlock,
+    toBlock,
+    userAddress,
+  }: GetEventsInput): Promise<MovementsByBlock[]> {
+    return this.helpers.withdrawals({
+      protocolToken: await this.getProtocolTokenByAddress(protocolTokenAddress),
+      filter: { fromBlock, toBlock, userAddress },
+    })
+  }
+
+  async getDeposits({
+    protocolTokenAddress,
+    fromBlock,
+    toBlock,
+    userAddress,
+  }: GetEventsInput): Promise<MovementsByBlock[]> {
+    return this.helpers.deposits({
+      protocolToken: await this.getProtocolTokenByAddress(protocolTokenAddress),
+      filter: { fromBlock, toBlock, userAddress },
+    })
+  }
+
+  async getTotalValueLocked({
+    protocolTokenAddresses,
+    blockNumber,
+  }: GetTotalValueLockedInput): Promise<ProtocolTokenTvl[]> {
+    const protocolTokens = await this.getProtocolTokens()
+
+    return await this.helpers.tvl({
+      protocolTokens,
+      filterProtocolTokenAddresses: protocolTokenAddresses,
+      blockNumber,
+    })
+  }
+
+  /**
+   * The weETH on L2 unwraps 1-to-1 to the weETH on mainnet,
+   * because it follows the "Omnichain Fungible Token" (OFT) standard from LayerZero.
+   *
+   * So we can use the same rate as the Mainnet-weETH to unwrap the L2-weETH to Mainnet-eETH, and then to ETH.
+   *
+   * @See https://medium.com/layerzero-official/explaining-the-oft-standard-310de5e84052
+   */
+  async unwrap({ blockNumber }: UnwrapInput): Promise<UnwrapExchangeRate> {
+    const [protocolToken] = await this.getProtocolTokens()
+    if (!protocolToken) throw new Error('No protocol token found')
+
+    const underlyingToken = (protocolToken.underlyingTokens || [])[0]
+    if (!underlyingToken) throw new Error('No protocol token found')
+
+    const useBlockNumber = blockNumber ?? (await this.provider.getBlockNumber())
+    const l2Block = await this.provider.getBlock(useBlockNumber)
+    if (!l2Block)
+      throw new Error(`No L2 block found with block number ${blockNumber}`)
+    if (!l2Block.date)
+      throw new Error(
+        `Block with number ${blockNumber} has no date, but it's needed to determine the unwrap rate`,
+      )
+
+    const mainnetProvider = this.helpers.allJsonRpcProviders[Chain.Ethereum]
+    const weETHContractMainnet = WeETH__factory.connect(
+      WE_ETH_ADDRESS_MAINNET,
+      mainnetProvider,
+    )
+
+    const sameDateMainnetBlockNumber = await getBlockByDate(
+      mainnetProvider,
+      l2Block.date,
+    )
```

Comment:
This function call will do an average of log(N*) calls to `provider.getBlock` sequentially (unfortunately there doesn't seem to be a `getBlockByTimestamp` available). We will need to check that is doesn't slow down calling `getPositions` for these two L2 chains a lot.

*: N is the number of blocks on mainnet

Reply:
Also nit, but can't we pass `l2Block.timestamp` instead of passing `date` and then converting it to a timestamp.

Reply:
Hum, I tend to consider the type `Date` as the first class citizen for function params as it has better devX than timestamp. Especially here `getBlockByDate` is a generic util that could be re-used in various places.